test/tests/serial/mail_tls_dhparams.bats

```diff
@@ -1,127 +1,76 @@
 
+  init_with_defaults
+  cp "${DH_PARAMS_CUSTOM}" "${TEST_TMP_CONFIG}/dhparams.pem"
   common_container_setup
-  should_have_valid_checksum "${DH_CUSTOM_CHECKSUM}"
-  should_emit_warning
-}
-
-@test "testing tls: DH Parameters - Custom [ONE_DIR=1]" {
-  # shellcheck disable=SC2030
-  PRIVATE_CONFIG=$(duplicate_config_for_container . mail_dhparams_custom_1)
 
-  cp "${DH_CUSTOM_PARAMS}" "${PRIVATE_CONFIG}/dhparams.pem"
-
-  common_container_setup
-  should_have_valid_checksum "${DH_CUSTOM_CHECKSUM}"
-  should_emit_warning
+  _should_match_service_copies "${DH_CHECKSUM_CUSTOM}"
+  _should_emit_warning
 }
 
-function common_container_setup() {
-  # shellcheck disable=SC2031
-  docker run -d --name mail_dhparams \
-    -v "${PRIVATE_CONFIG}:/tmp/docker-mailserver" \
-    -v "$(pwd)/test/test-files:/tmp/docker-mailserver-test:ro" \
-    -e ONE_DIR="${DMS_ONE_DIR}" \
-    -h mail.my-domain.com \
-    --tty \
-    "${NAME}"
-
-  wait_for_finished_setup_in_container mail_dhparams
+# Verify integrity of the default supplied DH Params (ffdhe4096)
+function _should_match_mozilla_copy() {
+  local DH_CHECKSUM=$1
+
+  # Verify the FFDHE params file has not been modified (equivalent to `target/shared/ffdhe4096.pem.sha512sum`):
+  # 716a462baecb43520fb1ba6f15d288ba8df4d612bf9d450474b4a1c745b64be01806e5ca4fb2151395fd4412a98831b77ea8dfd389fe54a9c768d170b9565a25
+  local DH_CHECKSUM_MOZILLA
+  DH_CHECKSUM_MOZILLA=$(curl https://ssl-config.mozilla.org/ffdhe4096.txt -s | sha512sum | awk '{print $1}')
+  assert_equal "${DH_CHECKSUM}" "${DH_CHECKSUM_MOZILLA}"
 }
 
-# Ensures the docker image services (Postfix and Dovecot) have the intended DH files
-function should_have_valid_checksum() {
+# Ensures the docker image services (Postfix and Dovecot) have the expected DH files:
+function _should_match_service_copies() {
   local DH_CHECKSUM=$1
 
-  local DH_CHECKSUM_DOVECOT
-  DH_CHECKSUM_DOVECOT=$(docker exec mail_dhparams sha512sum /etc/dovecot/dh.pem | awk '{print $1}')
-  assert_equal "${DH_CHECKSUM_DOVECOT}" "${DH_CHECKSUM}"
+  function __should_have_expected_checksum() {
+    _run_in_container bash -c "sha512sum ${1} | awk '{print \$1}'"
+    assert_success
+    assert_output "${DH_CHECKSUM}"
+  }
 
-  local DH_CHECKSUM_POSTFIX
-  DH_CHECKSUM_POSTFIX=$(docker exec mail_dhparams sha512sum /etc/postfix/dhparams.pem | awk '{print $1}')
-  assert_equal "${DH_CHECKSUM_POSTFIX}" "${DH_CHECKSUM}"
+  __should_have_expected_checksum '/etc/dovecot/dh.pem'
+  __should_have_expected_checksum '/etc/postfix/dhparams.pem'
 }
 
-function should_emit_warning() {
-  run sh -c "docker logs mail_dhparams | grep 'Using self-generated dhparams is considered insecure.'"
+function _should_emit_warning() {
+  run docker logs "${CONTAINER_NAME}"
   assert_success
+  assert_output --partial '[ WARNING ]  Using self-generated dhparams is considered insecure - unless you know what you are doing, please remove'
 }
```

Comment:
Useless function? This is called just once. Is it intended to be used in the future as well? If not, please move the inside of this function into the test.

Reply:
I tend to do this for tests sometimes so the test case itself is simpler to grok :sweat_smile: 

I'll make this change, but then I'm wondering if I should do the same for the mozilla copy method? Personally less noise in the test case is nicer. I don't see any harm with the approach, but happy to adjust :+1: (_I realize this one at least is pretty self-explanatory that it's a tad silly being extracted out to a function_ :sweat_smile: )

Reply:
I'd definitely go for it with `_should_emit_warning`, but probably also for `_should_match_mozilla_copy`. I'd leave `_should_match_service_copies` as is though :)